admin-api-server/lib/src/stm.ts

```diff
@@ -124,59 +118,103 @@ export class StateTransitionMachine {
+      this.#addVote(nft, actorId, attr, 'no');
+      this.#removeVote(nft, actorId, attr, 'yes');
+    } else {
+      throw `bad vote value ${v}`;
     }
-    nft.attributes[attr] = nft.attributes[attr].filter(
-      (id: number) => id !== actorId,
-    );
   }
 
-  #tryAttributeAddVote(nft: Nft, attr: string, actorId: number) {
+  #addVote(nft: Nft, actorId: number, attr: string, side: 'yes' | 'no') {
     if (!nft.attributes.hasOwnProperty(attr)) {
-      nft.attributes[attr] = [];
+      nft.attributes[attr] = {
+        yes: [],
+        no: [],
+      };
     }
-    if (nft.attributes[attr].some((id: number) => id === actorId)) {
-      throw `actor with id ${actorId} already signed nft with attr '${attr}', nft=${JSON.stringify(
+    if (nft.attributes[attr][side].some((id: number) => id === actorId)) {
+      throw `actor with id ${actorId} already voted ${side} nft with attr '${attr}', nft=${JSON.stringify(
         nft,
       )}'`;
     }
-    nft.attributes[attr].push(actorId);
+    nft.attributes[attr][side].push(actorId);
+  }
+
+  #removeVote(nft: Nft, actorId: number, attr: string, side: 'yes' | 'no') {
+    if (!nft.attributes.hasOwnProperty(attr)) {
+      return;
+    }
+    nft.attributes[attr][side] = nft.attributes[attr][side].filter(
+      (id: number) => id !== actorId,
+    );
   }
 
   // move nft if possible to a new state
-  // returns true if moved and adjusts nft in memory
-  // returns false if not moved
-  tryMoveNft(nft: Nft): boolean {
+  // returns for each possible transition target the list of conditions that
+  // were not met (and if we did transition, returns []).
+  tryMoveNft(nft: Nft): any {
     const st = this.states[nft.state];
 
-    for (const transition of st.transitions) {
-      if (evalExpr<boolean>(nft, transition.when, false)) {
+    const unmetTransitionConditions: any = {};
+    for (const transition of st?.transitions || []) {
+      let allSucceed = true;
+
+      unmetTransitionConditions[transition.next_state] = [];
+      for (const condition of transition.when) {
+        if (!evalExpr<boolean>(nft, condition, false)) {
+          allSucceed = false;
+
+          unmetTransitionConditions[transition.next_state].push(condition);
+        }
+      }
+
+      if (allSucceed) {
         log.notice(
           `nft.id ${nft.id} ${nft.state} -> ${
             transition.next_state
           }, attrs=${JSON.stringify(nft.attributes)}`,
         );
+
         nft.state = transition.next_state;
         if (typeof transition.do !== 'undefined') {
           execExpr(nft, transition.do);
         }
-        return true;
+        return {};
       }
     }
 
-    return false;
+    return unmetTransitionConditions;
   }
 }
+
+function isNone(v: any): boolean {
+  // note: v here is checked for both null and undefined
+  //       because undefined coerces into null. And its safe because nothing
+  //       else coerces into null (other than null itself).
+  return v == null;
+}
```

Comment:
Bravo !

Reply:
If the codeline itself makes little sense while reading it, can always put it into a helper function that is named properly haha